crates/milli/src/search/new/geo_sort.rs

```diff
@@ -267,24 +265,79 @@ impl<'ctx, Q: RankingRuleQueryTrait> RankingRule<'ctx, Q> for GeoSort<Q> {
                 cache.pop_back()
             }
         };
-        while let Some((id, point)) = next(&mut self.cached_sorted_docids) {
-            if geo_candidates.contains(id) {
-                return Ok(Some(RankingRuleOutput {
-                    query,
-                    candidates: RoaringBitmap::from_iter([id]),
-                    score: ScoreDetails::GeoSort(score_details::GeoSort {
-                        target_point: self.point,
-                        ascending: self.ascending,
-                        value: Some(point),
-                    }),
-                }));
+        let put_back = |cache: &mut VecDeque<_>, x: _| {
+            if ascending {
+                cache.push_front(x)
+            } else {
+                cache.push_back(x)
             }
-        }
+        };
 
-        // if we got out of this loop it means we've exhausted our cache.
-        // we need to refill it and run the function again.
-        self.fill_buffer(ctx, &geo_candidates)?;
-        self.next_bucket(ctx, logger, universe)
+        let mut current_bucket = RoaringBitmap::new();
+        // current_distance stores the first point and distance in current bucket
+        // The farthest distance between two points on earth is about 2e7 meters, u32 is big enough to hold any distance.
+        let mut current_distance: Option<([f64; 2], u32)> = None;
+        loop {
+            // The loop will only exit when we have found all points with equal distance or have exhausted the candidates.
+            if let Some((id, point)) = next(&mut self.cached_sorted_docids) {
+                if geo_candidates.contains(id) {
+                    let distance = distance_between_two_points(&self.point, &point).round() as u32;
+                    if let Some((point0, bucket_distance)) = current_distance.as_ref() {
+                        if bucket_distance != &distance {
+                            // different distance, point belongs to next bucket
+                            put_back(&mut self.cached_sorted_docids, (id, point));
+                            return Ok(Some(RankingRuleOutput {
+                                query,
+                                candidates: current_bucket,
+                                score: ScoreDetails::GeoSort(score_details::GeoSort {
+                                    target_point: self.point,
+                                    ascending: self.ascending,
+                                    value: Some(point0.to_owned()),
+                                }),
+                            }));
+                        } else {
+                            // same distance, point belongs to current bucket
+                            current_bucket.push(id);
+                            // remove from cadidates to prevent it from being added to the cache again
+                            geo_candidates.remove(id);
+                        }
```

Comment:
While chatting with the team, we thought about this else and the way it could loop through the whole dataset if all documents share the same distance.
As a security hatch, we think it would be great if you could exit the loop anyway once we reach a certain number of documents in the same bucket.

1000 seems like a good value, but it would be awesome if you could make the thresholds configurable so we can easily test it out

Reply:
I am worried that this will cause another problem. The ranking rules after geo need to get all the documents to sort correctly.If we only give 1000 documents, it may lead to wrong results.

For example, let's assume that all documents have the same distance, and the user sorts ["geoPoint...", "score:desc"]. Due to the security hatch, the document with the highest score may not be able to enter the bucket. in this case, the first item in the sorted result is not the document with the highest score.

Reply:
Yes, the idea is that;
- Generally speaking there is no reason to have more than a few documents exactly at the same distance, that’s really an edge case
- If a malevolent actor notices want to DOS an instance, they could push a ton of document exactly in the same spot and then every search requests would be irresponsive 

So, we think it’s better to get results slightly wrong in a case that should never happen that risk a DOS 😅